app/code/Magento/Checkout/Model/Cart.php

```diff
@@ -732,11 +742,6 @@ public function updateItem($itemId, $requestInfo = null, $updatingParams = null)
             throw new \Magento\Framework\Exception\LocalizedException(__($result));
         }
 
-        $this->_eventManager->dispatch(
-            'checkout_cart_product_update_after',
-            ['quote_item' => $result, 'product' => $product]
-        );
-        $this->_checkoutSession->setLastAddedProductId($productId);
```

Comment:
According to the logic above, if any exception is thrown, then the both `dispatch` and `setLastAddedProductId` are never called. I can not imagine any case which might lead to `Undefined variable` error related $product, $productId variables.

I agree that the `Magento\Checkout\Model\Cart::updateItem` method is unproperly built on exceptions and need to be refactored entirely. Feel free to fix it and cover the changes with an appropriate automated test.

Thank you!